File: 0725_leetcode.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        curr, length = head, 0
        while curr:
            curr, length = curr.next, length + 1

        chunk_size, longer_chunks = length // k, length % k
        res = [chunk_size + 1] * longer_chunks + [chunk_size] * (k - longer_chunks)

        prev, curr = None, head
        for index, num in enumerate(res):
            if prev:
                prev.next = None
            res[index] = curr
            for i in range(num):
                prev, curr = curr, curr.next
        return res

File: 0725_leetcode.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getLength(head *ListNode) int {
	length := 0
	for head != nil {
		length++
		head = head.Next
	}
	return length
}

func getList(curr *ListNode, head *ListNode, quo, rem int) (*ListNode, *ListNode) {
	if head == nil {
		return nil, nil
	}
	l := 0
	l += quo
	if rem != 0 {
		l += 1
	}
	ref := head
	prev := new(ListNode)
	prev = nil
	for i := 0; i < l; i++ {
		prev = head
		head = head.Next
	}
	curr = ref
	prev.Next = nil
	return curr, head
}

func splitListToParts(head *ListNode, k int) []*ListNode {
	ans := make([]*ListNode, 0)
	n := getLength(head)
	quo := n / k
	rem := n % k
	for i := 0; i < k; i++ {
		curr := new(ListNode)
		curr = nil
		curr, head = getList(curr, head, quo, rem)
		if rem != 0 {
			rem--
		}
		ans = append(ans, curr)
	}
	return ans
}